// colors/lib/custom/trap-compiled.js
"use strict";

module["exports"] = function runTheTrap(text, options) {
  var result = "";
  text = text || "Run the trap, drop the bass";
  text = text.split("");
  var trap = {
    a: ["@", "Ą", "Ⱥ", "Ʌ", "Δ", "Λ", "Д"],
    b: ["ß", "Ɓ", "Ƀ", "ɮ", "β", "฿"],
    c: ["©", "Ȼ", "Ͼ"],
    d: ["Ð", "Ɗ", "Ԁ", "ԁ", "Ԃ", "ԃ"],
    e: ["Ë", "ĕ", "Ǝ", "ɘ", "Σ", "ξ", "Ҽ", "੬"],
    f: ["Ӻ"],
    g: ["ɢ"],
    h: ["Ħ", "ƕ", "Ң", "Һ", "Ӈ", "Ԋ"],
    i: ["༏"],
    j: ["Ĵ"],
    k: ["ĸ", "Ҡ", "Ӄ", "Ԟ"],
    l: ["Ĺ"],
    m: ["ʍ", "Ӎ", "ӎ", "Ԡ", "ԡ", "൩"],
    n: ["Ñ", "ŋ", "Ɲ", "Ͷ", "Π", "Ҋ"],
    o: ["Ø", "õ", "ø", "Ǿ", "ʘ", "Ѻ", "ם", "۝", "๏"],
    p: ["Ƿ", "Ҏ"],
    q: ["্"],
    r: ["®", "Ʀ", "Ȑ", "Ɍ", "ʀ", "Я"],
    s: ["§", "Ϟ", "ϟ", "Ϩ"],
    t: ["Ł", "Ŧ", "ͳ"],
    u: ["Ʊ", "Ս"],
    v: ["ט"],
    w: ["Ш", "Ѡ", "Ѽ", "൰"],
    x: ["Ҳ", "Ӿ", "Ӽ", "ӽ"],
    y: ["¥", "Ұ", "Ӌ"],
    z: ["Ƶ", "ɀ"]
  };
  text.forEach(function (c) {
    c = c.toLowerCase();
    var chars = trap[c] || [" "];
    var rand = Math.floor(Math.random() * chars.length);
    if (typeof trap[c] !== "undefined") {
      result += trap[c][rand];
    } else {
      result += c;
    }
  });
  return result;
};
